src/qibochem/ansatz/hardware_efficient.py

```diff
@@ -1,26 +1,17 @@
 from qibo import gates
 
 
-def hea_su2(nlayers, nqubits):
-    """
-    implements a hardware-efficient ansatz SU2
-    inputs:
-    """
+def hea(nlayers, nqubits, parameter_gates=["RY", "RZ"], coupling_gates="CZ"):
     gate_list = []
 
     for ilayer in range(nlayers):
-        for iqubit in range(nqubits):
-            gate_list.append(gates.RY(iqubit, theta=0))
-        for iqubit in range(nqubits):
-            gate_list.append(gates.RZ(iqubit, theta=0))
+        for igate in range(len(parameter_gates)):
+            for iqubit in range(nqubits):
+                gate_list.append(getattr(gates, parameter_gates[igate])(iqubit, theta=0))
+
         # entanglement
         for iqubit in range(nqubits - 1):
-            gate_list.append(gates.CNOT(iqubit, iqubit + 1))
-        gate_list.append(gates.CNOT(nqubits - 1, 0))
-
-    for iqubit in range(nqubits):
-        gate_list.append(gates.RY(iqubit, theta=0))
-    for iqubit in range(nqubits):
-        gate_list.append(gates.RZ(iqubit, theta=0))

```

Comment:
Isn't the final layer of rotation gates still necessary? If so, just add the (edited) code from above:
```
    for r_gate in parameter_gates:
        for iqubit in range(nqubits):
            gate_list.append(getattr(gates, r_gate)(iqubit, theta=0))
```

Random thought: Maybe can add a function parameter to control if the final layer is added or not

Reply:
Some sources include it:
https://www.nature.com/articles/nature23879

Some don't:
https://www.classiq.io/insights/generating-a-hardware-efficient-ansatz

I didn't because I was lazy to include more terms in parameter count. 

But perhaps, it is better to think of it as adding an initial layer of ry/rz than each subsequent layer is made of coupling and parameter gates. Maybe we should include this? But there should also be some easy way to correctly count the number of parameter gates for parameter count later, maybe as a feature? This we can think about and discuss on implementation.

Reply:
My personal experience is that the results with a final layer are quite a bit better than those without, but I never really did test it thoroughly. So, I'm actually OK either way, but we can just keep the current implementation for now.

Re: number of circuit parameters. If the ansatz is just used as a standalone circuit, the easy way would be to just call something like `len(circuit.get_parameters())` or `len(circuit.parametrized_gates)`, and don't bother thinking about how to count the number of parametrized gates😬

Reply:
Re: number of circuit parameters
Did some quick tests, `circuit.get_parameters()` gives only the parameters without the gate info, `circuit.parametrized_gates()` gives the parameterized gate but in memory, i.e. returns something like `[<qibo.gates.gates.RY object at 0x7f5a3bf25fd0>, <qibo.gates.gates.RX object at 0x7f5a25fa4e20>]`, however you can still directly access the gate properties through looping over the list and using `gate.name', `gate.parameters`, `gate.target_qubits`, e.t.c to reconstruct a list of parameterized gates and whatever info you'll need.